### ISSUE TITLE
Vibrations when using velocity interface via ros_control

### ISSUE THREAD

**Issue by miguelprada** [Zagitta/ur_modern_driver#15]
I finally got to spend some time testing this version of the driver and, to my despair, initial results are not very promising for me... :sweat: 

I found that moving my UR5 using `ros_control` and the velocity interface results in very noisy movements due to high frequency vibrations. The exact same setup but substituting master in this repo for master from ThomasTimm works just fine.

The simplest possible setup I could come up with to reproduce my issues is launching the robot with the default `ur5_ros_control` launchfile and commanding simple, single joint motion using `rqt_joint_trajectory_controller`. Specifically I send the elbow joint from 0 to pi/2 in a single command, letting the joint trajectory controller interpolate. For this I input the desired position in the text box, to avoid variations that would be introduced by the sliders.

Plots below compare desired and actual positions and velocities as reported by the joint trajectory controller, both for the master version in this repo and the master version in ThomasTimm's.

From **zagitta/master**:

![zagitta/master position](https://user-images.githubusercontent.com/678580/35927946-3fa34376-0c2c-11e8-9bda-80dbfda6d61d.png)

![zagitta/master velocity](https://user-images.githubusercontent.com/678580/35927986-54dd2810-0c2c-11e8-9742-5c560324e3ad.png)

From **thomastimm/master**:

![thomastimm/master position](https://user-images.githubusercontent.com/678580/35928080-91eddb64-0c2c-11e8-8b81-051004c5d952.png)

![thomastimm/master velocity](https://user-images.githubusercontent.com/678580/35928098-9b23782e-0c2c-11e8-937d-2cade95206c7.png)

To be clear. I'm not particularly interested in the tracking error here. That's a completely different matter. What I am interested in is the noisy velocity readings that are only appearing on this version of the driver.

I'll try to look a bit deeper into this, but I need to get familiar with the new driver implementation first, since it is much more difficult for me to follow than the previous one.

For completeness sake, I'm running my tests on ROS Indigo on Ubuntu Trusty and the specific commits I'm using are 0611afc and thomastimm/ur_modern_driver@9397f47. The robot is a UR5 attached to a CB3.1 running PolyScope 3.3.4.310. Also, the robot and the CB are directly connected to each other, with nothing in between.

I observed the same behavior for 0611afc with ROS Kinetic on Xenial, but could not do the comparison using that setup.

**Comment by miguelprada** [comment]
Since the `joint_trajectory_controller` in it's default configuration for the UR contains a [feedback loop](https://github.com/ros-controls/ros_controllers/blob/e5a6c5a5f9f84756b6601961a5b03b7dffc5118c/joint_trajectory_controller/include/joint_trajectory_controller/hardware_interface_adapter.h#L211), I run an additional test in pure feedforward control to better pinpoint where the vibrations are coming from.

**TL;DR;** The vibrations are still present so there must be something wrong in the _command_ side of the `ros_control` loop.

For this I repeated the same experiment as above, but using the recently introduced [velocity feedforward](https://github.com/ros-controls/ros_controllers/pull/227) term in `joint_trajectory_controller` and set all the PID (feedback) gains to 0.0. The idea was that if the motion was smooth in this case, then somehow the vibrations could be caused by noisy position and/or velocity feedback injected into the controllers.

The vibration can still be seen in the desired/actual velocity plot below. It could be argued that this plot is not necessarily representative, since the vibrations might only be part of the _feedback_ signals and therefore appear in the plot but not be present on the robot. However this is not the case, since the vibrations are clearly audible while the motion is executed, which is not the case with thomastimm/ur_modern_driver@9397f47.

![z_master_vel_ff_only](https://user-images.githubusercontent.com/678580/36022358-238da308-0d89-11e8-944e-38e417dbcb48.png)

To reproduce with ROS Indigo on Trusty, use the following versions:

```yaml
- git:
    local-name: ros_controllers
    uri: https://github.com/miguelprada/ros_controllers.git
    version: 773b2c0
- git:
    local-name: ur_modern_driver
    uri: https://github.com/miguelprada/ur_modern_driver.git
    version: 8cc8f93
```

Start the robot with `roslaunch ur_modern_driver ur5_ros_control.launch robot_ip:=RO.BO.T.IP` and move joints around using the `rqt_joint_trajectory_controller` (or any other method of your choice).

It should be noted that the position tracking with this setup is terrible, because of the pure velocity feedforward with no PID closing the position control loop. Again, it is not the point being treated here.

**Comment by miguelprada** [comment]
For comparison's sake, the same experiment based on thomastimm/ur_modern_driver@9397f47 produces the following output.

![tt_master_vel_ff_only](https://user-images.githubusercontent.com/678580/36024557-5b0a26f6-0d90-11e8-89a2-554f395e6a4c.png)

Use these to reproduce:

```yaml
- git:
    local-name: ros_controllers
    uri: https://github.com/miguelprada/ros_controllers.git
    version: 773b2c0
- git:
    local-name: ur_modern_driver
    uri: https://github.com/miguelprada/ur_modern_driver.git
    version: 5d54cd3
```

**Comment by gavanderhoorn** [comment]
Thanks for the analysis, even if it's slightly worrying.

I can't be of much concrete help right now, but perhaps something to check is whether the output of the velocity controller actually makes sense? If it does, then it could be an issue with the hw iface implementation.

**Comment by miguelprada** [comment]
> Thanks for the analysis, even if it's slightly worrying.

Worrying indeed...

> I can't be of much concrete help right now, but perhaps something to check is whether the output of the velocity controller actually makes sense? If it does, then it could be an issue with the hw iface implementation.

With the pure feedforward setup in the latest experiments, the output of the controller is exactly what is plotted in the purple lines.

I believe this must have something to do with the implementation of the ros_control loop in this version of the driver, but as I already said before, its difficult for me to say what. Do you think a wireshark capture of the data flow between the PC and the UR controller might be of help?

**Comment by gavanderhoorn** [comment]
> With the pure feedforward setup in the latest experiments, the output of the controller is exactly what is plotted in the purple lines.

I was actually wondering whether there is something 'happening' between where the controller writes new values to the hw iface and where the hw iface (in the driver) forwards them.

The velocity interface is implemented in [these lines](https://github.com/Zagitta/ur_modern_driver/blob/0611afcf40b494e39ab418eebb9c1700d14fbac6/src/ros/hardware_interface.cpp#L33-L63). Perhaps the clamping is not implemented correctly? I see a lot of oscilation in your plots.

> Do you think a wireshark capture of the data flow between the PC and the UR controller might be of help?

hm, not sure. I'd first take a look at the code.

**Comment by gavanderhoorn** [comment]
What do you have your `CMAKE_BUILD_TYPE` set to? It's probably unrelated, but there's quite a difference in CPU usage between `Debug` or anything with optimisations with this fork of `ur_modern_driver`. It shouldn't affect any of the things that you are reporting, but might be something to check.

**Comment by miguelprada** [comment]
> > Do you think a wireshark capture of the data flow between the PC and the UR controller might be of help?

> hm, not sure. I'd first take a look at the code.

:+1:

> What do you have your CMAKE_BUILD_TYPE set to? It's probably unrelated, but there's quite a difference in CPU usage between Debug or anything with optimisations with this fork of ur_modern_driver. It shouldn't affect any of the things that you are reporting, but might be something to check.

I don't set it explicitly, so whatever it's default value is. And in any case it should be the same in both cases, with code from `zagitta/master` and from `ThomasTimm/master`.

**Comment by gavanderhoorn** [comment]
> > What do you have your CMAKE_BUILD_TYPE set to? It's probably unrelated, but there's quite a difference in CPU usage between Debug or anything with optimisations with this fork of ur_modern_driver. It shouldn't affect any of the things that you are reporting, but might be something to check.
> 
> I don't set it explicitly, so whatever it's default value is. And in any case it should be the same in both cases, with code from `zagitta/master` and from `ThomasTimm/master`.

true, but I think you're only considering functional aspects.

If optimisation leads to lower cpu usage, and thus shorter execution times, it could influence some non-functional aspect that in the end causes what you describe in your OP.

But as I wrote earlier: I'd first check the code.

**Comment by miguelprada** [comment]
I think thomastimm/ur_modern_driver#92 is exactly what's happening to me. Worst thing is I already faced this issue and [patched it](https://github.com/ThomasTimm/ur_modern_driver/issues/100) before. :man_facepalming: 

So far I only tested it against the UR simulator. If I'm able to replicate on the real robot, once I have access to it again, I'll prepare a PR adding a new [`URCommander`](https://github.com/Zagitta/ur_modern_driver/blob/master/include/ur_modern_driver/ur/commander.h) subclass.

**Comment by gavanderhoorn** [comment]
Bah. I hate these things.

💯 for digging into this.
